### ISSUE TITLE
`torch.cuda.amp.GradScaler` may skip parameter synchronization required by post localSGD optimizer

### ISSUE THREAD

**Issue by wayi1** [pytorch/pytorch#76282]
### 🐛 Describe the bug

The following two designs may not work very well together.

-  When mixed precision is enabled by `torch.cuda.amp.GradScaler`, as noted in this [line](https://github.com/pytorch/pytorch/blob/bbc3cc6718d841665ab2677477d458ed97b75609/torch/cuda/amp/grad_scaler.py#L68) and this [line](https://github.com/pytorch/pytorch/blob/bbc3cc6718d841665ab2677477d458ed97b75609/torch/cuda/amp/grad_scaler.py#L86), if gradients contain infs/NaNs, then `optimizer.step()` can be skipped.
- Some [distributed optimizers](https://github.com/pytorch/pytorch/tree/77f23d64607142c0cba47f5148ebc1959c9de366/torch/distributed/optim) often run certain communications besides `step()` of the local optimizer.

A prominent example is [post_localSGD_optimizer](https://github.com/pytorch/pytorch/blob/77f23d64607142c0cba47f5148ebc1959c9de366/torch/distributed/optim/post_localSGD_optimizer.py#L78), which runs parameter averaging after `step()` of local optimizer. Since parameter/model averaging paradigm (e.g., local SGD, hierarchical SGD, SlowMo) does not synchronize parameters at every iteration, even if gradients have infs/NaNs and the local trainable parameters are not updated at the current iteration, it is still better to average the parameters than have such averaging skipped. Otherwise the parameters might become too stale (e.g., the last synchronization could be `2*T` iterations ago, where `T` denotes the averaging period). IMO, this might even cause the loss of NaN might occur for some cases.

Furthermore, this behavior of silently skipping `optimizer.step()` can also be a risk for some other (future) distributed optimizers if certain states (e.g., optimizer state, buffer) still need to be updated. For example, the internal step counter [here](https://github.com/pytorch/pytorch/blob/77f23d64607142c0cba47f5148ebc1959c9de366/torch/distributed/optim/functional_rmsprop.py#L91) can be less than the # of steps in a training loop, and same as the step counter in this [RFC](https://github.com/pytorch/pytorch/issues/74556). Additionally, not sure if any other optimizer wrapper may need to run certain communications even if no local optimizer step.

One relevant feature request is https://github.com/pytorch/pytorch/issues/67590, which proposes to signal that the scaler has been stepped and `optimizer.step()` has been actually called.

_Note that this issue may not be very serious for some cases in practice, because [such step skipping is expected to occur rarely after the first few iterations](https://github.com/pytorch/pytorch/blob/bbc3cc6718d841665ab2677477d458ed97b75609/torch/cuda/amp/grad_scaler.py#L93-L95)._

### Versions

PyTorch 1.11
